chainstate/src/detail/tests/normal_tokens.rs

```diff
@@ -0,0 +1,843 @@
+
+        // Now D2 trying to spend tokens from mainchain
+        let values = vec![
+            OutputValue::Token(TokenData::TokenTransferV1 {
+                token_id,
+                amount: ISSUED_FUNDS,
+            }),
+            OutputValue::Coin(Amount::from_atoms(123454)),
+        ];
+        let (block_d2, block_index) = process_token_ex(
+            &mut test_framework,
+            vec![spent_input],
+            ParentBlock::BlockId(Id::<GenBlock>::from(block_c2.get_id())),
+            values,
+        )
+        .unwrap();
+        assert!(block_index.is_none(), "Reog is not allowed at this height");
+
+        // Block E2 will cause reorganization
+        let values = vec![OutputValue::Coin(Amount::from_atoms(123453))];
+        assert!(matches!(
+            process_token_ex(
+                &mut test_framework,
+                vec![],
+                ParentBlock::BlockId(Id::<GenBlock>::from(block_d2.get_id())),
+                values
+            ),
+            Err(BlockError::StateUpdateFailed(
+                ConnectTransactionError::MissingOutputOrSpent
+            ))
+        ));
+    })
+}
+
+#[test]
+fn test_attempt_to_print_tokens() {
+    utils::concurrency::model(|| {
+        const ISSUED_FUNDS: Amount = Amount::from_atoms(987_654_321);
+
+        // Issue a new token
+        let mut test_framework = TestFramework::default();
+        let values = vec![
+            OutputValue::Token(TokenData::TokenIssuanceV1 {
+                token_ticker: b"SOME".to_vec(),
+                amount_to_issue: ISSUED_FUNDS,
+                number_of_decimals: 1,
+                metadata_uri: b"https://some_site.meta".to_vec(),
+            }),
+            OutputValue::Coin(Amount::from_atoms(123456)),
+        ];
+        let (block_a, _) = process_token_ex(
+            &mut test_framework,
+            vec![],
+            ParentBlock::BestBlock,
+            values.clone(),
+        )
+        .unwrap();
+        assert_eq!(block_a.transactions()[0].outputs()[0].value(), &values[0]);
+        let token_id = token_id(&block_a.transactions()[0]).unwrap();
+
+        // Try to transfer a bunch of outputs where each separately do not exceed input tokens value, but a sum of outputs larger than inputs.
+        let values = vec![
+            OutputValue::Token(TokenData::TokenTransferV1 {
+                token_id,
+                amount: ISSUED_FUNDS,
+            }),
+            OutputValue::Token(TokenData::TokenTransferV1 {
+                token_id,
+                amount: ISSUED_FUNDS,
+            }),
+            OutputValue::Coin(Amount::from_atoms(123454)),
+        ];
+
+        assert!(matches!(
+            process_token(&mut test_framework, ParentBlock::BestBlock, values),
+            Err(BlockError::StateUpdateFailed(
+                ConnectTransactionError::AttemptToPrintMoney(_, _)
+            ))
+        ));
+    });
+}
+
+#[test]
+fn test_attempt_to_mix_input_tokens() {
```

Comment:
What does this test exactly? Can you elaborate, plase?

Reply:
In a few words, we issue tokens A and tokens B, and try to spend tokens A with values A+B. If tokens values calculate correctly in connect block, it will fail with `AttemtToPrintMoney`.